### ISSUE TITLE
What is the difference from id_tree?

### ISSUE THREAD

**Issue by xiaoniu-578fa6bff964d005** [iwburns/slab-tree#20]
According to the doc, these two crates have pretty similar project goals.

Can we have more detailed comparison between them in the crate description?

**Comment by iwburns** [comment]
Thanks for your interest in `slab_tree` (and `id_tree`)!

I would first like to make it clear that I wouldn't consider this project (`slab_tree`) to be "ready" for real-world use yet.  It certainly _could_ be used in "real" projects and I don't know of any reason why someone shouldn't do so; however, I'm not quite finished with the design so things could change between now and `1.0.0`.

You're correct, these two crates are quite similar in their goals, but there are a few differences in approach between the two.

1. `id_tree` uses `Vec`s to store each node's children instead of just storing five `Option<NodeId>`s (parent, first/last child, next/prev sibling) like `slab_tree` does.
1. `id_tree` uses a `Vec` to store free indexes so that spaces inside the slab can be re-used.  `slab_tree` uses generational indexes and "filled" / "empty" slots to store that information.
1. `id_tree` forces you to talk directly to the `Tree` object for all tree alterations, `slab_tree` allows you to modify the tree by calling methods on `NodeMut`s instead if you wish.

There are probably other differences between the two crates, but I think this covers all of the large ones (at the current time).  The differences above likely have performance implications, but I haven't run any benchmarks yet to determine how big the differences might be.

I will certainly consider writing up a compare / contrast section on one of these crates at some point, but I don't think `slab_tree` is solid enough in its design for me to give any more of a write-up than what you see above.

**Comment by clbarnes** [comment]
Were those differences made with the intention of improving performance or are they just design choices which could improve or worsen things?

**Comment by bjadamson** [comment]
Same. I came across these two crates and I'm having trouble figuring out which one I should invest time into learning / submitting PR's to.

Thanks for putting your time into developing these crates BTW (:

**Comment by iwburns** [comment]
@clbarnes (sorry for the delayed response here)

>Were those differences made with the intention of improving performance or are they just design choices which could improve or worsen things?

They are simply design choices that I was trying to explore. When I wrote `id_tree` I wasn't sure which choice would be better, so I just went with one and tried it out. I went down that path for some time, developing `id_tree` and not really considering the other approach for a while. Some time later the thought occurred to me to try out the other approach so I decided to give it a swing with `slab_tree`. I'd like to be clear though, I had some vague notion that they might behave differently in terms of performance, but I haven't done any testing to see what the differences are in reality. The exploration I was doing at the time was more from an implementation/API perspective than a performance perspective.

@bjadamson 
>I'm having trouble figuring out which one I should invest time into learning / submitting PR's to.

I don't know if I can give you a definitive answer here, but I'll give you what I can:
- I currently consider `id_tree` to be the more stable of the two (as it is in v1.x.x territory).
- I personally prefer the implementation and API of `slab_tree` over the implementation of `id_tree`, but it probably has fewer features than `id_tree`.

To be honest, I'm not exactly "happy" with the current state of either of these projects and I feel like they could both be improved. I don't have a list of issues enumerated anywhere; these are just frustrations I've had while working on them and trying to come up with the right design for each approach.

These projects are currently both in "passive maintenance" mode from my perspective. If people are interested in contributing, I'd certainly be happy to discuss design ideas, give/receive feedback, and provide implementation guidance wherever necessary. Please (anyone) let me know if you're interested in this.

>Thanks for putting your time into developing these crates BTW (:

That's very nice of you to say! I'd like to thank you and anyone else reading here for you interest in these projects as well. Both projects started out as proofs of concept / experimental ideas that I wanted to play with; so it's nice to hear that they might be helpful to others as well.

**Comment by clbarnes** [comment]
Thank you for the information! I did go with `slab_tree` in the end for [my crate](https://github.com/clbarnes/neurarbor).

**Comment by bjadamson** [comment]
> That's very nice of you to say! I'd like to thank you and anyone else reading here for you interest in these projects as well. Both projects started out as proofs of concept / experimental ideas that I wanted to play with; so it's nice to hear that they might be helpful to others as well.

Your welcome. I'm trying to see if either (at this point after reading your response I'm leaning towards slab-tree) libraries can be used to build a dynamic and branching algorithm for NPC dialogue in a game I am working on.

> * I currently consider id_tree to be the more stable of the two (as it is in v1.x.x territory).
> * I personally prefer the implementation and API of slab_tree over the implementation of id_tree, but it probably has fewer features than id_tree.

Thanks for elaborating here.

> These projects are currently both in "passive maintenance" mode from my perspective. If people are interested in contributing, I'd certainly be happy to discuss design ideas, give/receive feedback, and provide implementation guidance wherever necessary. Please (anyone) let me know if you're interested in this.

Do you think the use cases could potentially be merged? From what I've learned so far it seems as if slab-tree is a different implementation strategy from id-tree and the interface is very-similar from the consumer's pov.

**Comment by spearman** [comment]
FWIW i ended up sticking with id_tree as it performed better in benchmarks compared to slab-tree and ego-tree:
```
test tests::bench_build_ego_tree                 ... bench:   7,618,563 ns/iter (+/- 13,729)
test tests::bench_build_id_tree                  ... bench:   4,924,486 ns/iter (+/- 35,088)
test tests::bench_build_slab_tree                ... bench:   9,609,665 ns/iter (+/- 36,074)
test tests::bench_combined_ego_tree              ... bench:   7,493,492 ns/iter (+/- 21,162)
test tests::bench_combined_ego_tree_insert_order ... bench:   1,188,620 ns/iter (+/- 2,288)
test tests::bench_combined_id_tree               ... bench:   5,149,779 ns/iter (+/- 28,222)
test tests::bench_combined_slab_tree             ... bench:   9,626,087 ns/iter (+/- 52,441)
test tests::bench_traverse_ego_tree              ... bench:      29,619 ns/iter (+/- 56)
test tests::bench_traverse_id_tree               ... bench:      15,757 ns/iter (+/- 162)
test tests::bench_traverse_slab_tree             ... bench:      42,258 ns/iter (+/- 205)
```
https://gist.github.com/spearman/61f3ba951d122a53687db5ad556752e2

I'm using id_tree in a GUI library as the main tree structure (not yet released, but will be open sourced after i write an opengl backend), works well for my purposes, haven't had any problems.

**Comment by iwburns** [comment]
>Do you think the use cases could potentially be merged? [...] the interface is very-similar from the consumer's pov.

You're right that they are similar, but there is one slightly larger difference between the two from the consumer's perspective:

1. in `id_tree` if you need to do an insertion or a removal, you _must_ talk to the `Tree` itself via `tree.insert(...)` or `tree.remove(...)`.
2. in `slab_tree` there is no `tree.insert(...)` at all, only a `tree.remove(...)`. Instead of inserting via the `Tree` you talk to a `NodeMut` to do insertions via `node_mut.append(...)` or `node_mut.prepend(...)`.  `NodeMut` also supports removal via `node_mut.remove_first(...)` and `node_mut.remove_last(...)`.

Otherwise, I believe the two interfaces are _mostly_ quite similar to one another.

As for whether or not they could be merged:

I think they _could_, but I don't know if they _should_. It's certainly something I've thought about before, but I've made no moves in that area. If that happened, there would be some design work needed to figure out which parts of the public API make sense to keep and which make sense to remove.

There's also the matter of picking a "back-end" to use in the merged version. `slab_tree` and `id_tree` both roll their own slab-type structures to store the data, but they're slightly different from one another.  I _think_ I prefer `slab_tree`'s implementation, but it would also be good to explore other crates for this purpose. I know there are several slab/arena storage crates out there that might be worth looking at, but I haven't done that research in quite a while so the landscape has probably changed a bit since the last time I looked.

@spearman thanks for the info here.  That's a nice data-point to have!
